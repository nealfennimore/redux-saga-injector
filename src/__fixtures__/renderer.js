import PropTypes from 'prop-types';
import React from 'react';
import ReactDOMServer from 'react-dom/server';

import { Provider } from 'react-redux';
import createStore from '../__fixtures__/store';
import { preloadQueue } from '../sagas/queue';

const Application = ( {
    store,
    children,
} )=> (
    <Provider store={store}>
        { children }
    </Provider>
);

Application.propTypes = {
    store: PropTypes.shape( {} ).isRequired,
    children: PropTypes.node.isRequired,
};


const renderMarkup = ( App, {store} )=> (
    ReactDOMServer.renderToString(
        <Application store={store}>
            <App store={store} />
        </Application>
    )
);

export default function render( App, store, req, res, options ) {
    return async function() {
        // Server saga listens for any injected sagas to finish
        const preload = store.runSaga( preloadQueue, options );

        // Start initial render to start sagas
        // This is a throw away render
        renderMarkup( App, { store } );

        // Proceed after preload saga finishes
        await preload.done;

        // Get markup with updated store
        const html = renderMarkup( App, { store } );

        res.send( `
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id="app">${html}</div>
                </body>
            </html>
        ` );
    };
}
