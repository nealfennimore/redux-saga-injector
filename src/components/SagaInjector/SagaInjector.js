import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { cancelSagas, runSagas } from 'src/actions/sagas';
import { curry } from 'src/utils/fp';
import { genUID } from 'src/utils/uid';

/**
 * Inject sagas to run with component lifecycle
 *
 * @param {Object} options Should contain sagas as an array
 * @param {any} WrappedComponent Component to wrap
 * @returns SagaInjector
 */
function SagaInjectorHOC( options, WrappedComponent ) {
    const { sagas, STORE_KEY } = Object.assign( { sagas: [], STORE_KEY: 'store' }, options );

    const componentName = WrappedComponent.displayName ||
	WrappedComponent.name ||
    'Component';

    class SagaInjector extends Component {
        static displayName = `SagaInjector(${ componentName })`
        static propTypes = {
            [ STORE_KEY ]: PropTypes.object // eslint-disable-line
        }
        static contextTypes = {
            [ STORE_KEY ]: PropTypes.object
        }

        state = {
            injectedSagas: false,
        }

        uid = genUID()
        sagas = sagas

        componentWillMount() {
            this.injectSagas();
        }

        componentWillUnmount() {
            this.store.dispatch( cancelSagas( this.uid ) );
        }

        get store() {
            return this.props[STORE_KEY] || this.context[STORE_KEY];
        }

        get hasSagas() {
            return !! this.sagas.length;
        }

        get hasInjectedSagas() {
            return this.state.injectedSagas;
        }

        set hasInjectedSagas( injectedSagas ) {
            this.setState( {injectedSagas} );
        }

        injectSagas() {
            if ( this.hasSagas && ! this.hasInjectedSagas ) {
                this.store.dispatch( runSagas( this.sagas, this.uid ) );
                this.hasInjectedSagas = true;
            }
        }

        render() {
            return (
                <WrappedComponent {...this.props} />
            );
        }
    }

    return SagaInjector;
}

export default curry( SagaInjectorHOC );
