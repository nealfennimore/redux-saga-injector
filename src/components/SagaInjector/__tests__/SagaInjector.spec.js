import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import SagaInjector from '../SagaInjector';

jest.mock( 'src/utils/uid', ()=>( {
    genUID: ()=>'uid'
} ) );

class Test extends Component {
    render() {
        return 'test';
    }
}

describe( 'SagaInjector', ()=>{
    let store, sagas, TestComponent;
    beforeEach( ()=>{
        store = {
            dispatch: jest.fn()
        };

        sagas = [
            jest.fn(),
            jest.fn(),
            jest.fn()
        ];

        TestComponent = SagaInjector( { sagas } )( Test );
    } );

    test( 'should match snapshot', ()=>{
        const component = renderer.create(
            <TestComponent store={store} />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );

    describe( 'lifecycle', ()=>{
        test( 'should inject sagas on mount', ()=>{
            const component = mount(
                <TestComponent store={store} />
            );

            expect( component.instance().store ).toEqual( store );
            expect( component.instance().sagas ).toEqual( sagas );
            expect( component.instance().hasSagas ).toEqual( true );
            expect( store.dispatch.mock.calls ).toMatchSnapshot();
        } );
        test( 'should not inject when no sagas', ()=>{
            TestComponent = SagaInjector( {} )( Test );
            const component = mount(
                <TestComponent store={store} />
            );

            expect( component.instance().store ).toEqual( store );
            expect( component.instance().sagas ).toEqual( [] );
            expect( component.instance().hasSagas ).toEqual( false );
            expect( store.dispatch ).not.toBeCalled();
        } );
        test( 'should unmount properly', ()=>{
            const component = mount(
                <TestComponent store={store} />
            );
            const spy = jest.spyOn( component.instance(), 'componentWillUnmount' );
            component.update();
            component.unmount();
            expect( spy ).toBeCalled();
            expect( store.dispatch.mock.calls ).toMatchSnapshot();
        } );
    } );


} );
