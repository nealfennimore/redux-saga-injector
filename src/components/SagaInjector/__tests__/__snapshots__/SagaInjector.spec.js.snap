// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaInjector lifecycle should inject sagas on mount 1`] = `
Array [
  Array [
    Object {
      "sagas": Array [
        [MockFunction],
        [MockFunction],
        [MockFunction],
      ],
      "type": "@@redux-saga-injector/RUN_SAGAS",
      "uid": "uid",
    },
  ],
]
`;

exports[`SagaInjector lifecycle should unmount properly 1`] = `
Array [
  Array [
    Object {
      "sagas": Array [
        [MockFunction],
        [MockFunction],
        [MockFunction],
      ],
      "type": "@@redux-saga-injector/RUN_SAGAS",
      "uid": "uid",
    },
  ],
  Array [
    Object {
      "type": "@@redux-saga-injector/CANCEL_SAGAS",
      "uid": "uid",
    },
  ],
]
`;

exports[`SagaInjector should match snapshot 1`] = `"test"`;
